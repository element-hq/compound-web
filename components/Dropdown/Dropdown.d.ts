import { default as React } from '../../../node_modules/react';
type DropdownProps = {
    /**
     * The CSS class name.
     */
    className?: string;
    /**
     * The controlled value of the dropdown.
     */
    value?: string;
    /**
     * The default value of the dropdown, used when uncontrolled.
     */
    defaultValue?: string;
    /**
     * The values of the dropdown.
     * [value, text]
     */
    values: [string, string][];
    /**
     * The placeholder text.
     */
    placeholder: string;
    /**
     * The label to display at the top of the dropdown
     */
    label: string;
    /**
     * The help label to display at the bottom of the dropdown
     */
    helpLabel?: string;
    /**
     * Callback for when the value changes.
     * @param value
     */
    onValueChange?: (value: string) => void;
    /**
     * The error message to display.
     */
    error?: string;
};
/**
 * The dropdown content.
 */
export declare const Dropdown: React.ForwardRefExoticComponent<DropdownProps & React.RefAttributes<HTMLButtonElement>>;
export {};
//# sourceMappingURL=Dropdown.d.ts.map