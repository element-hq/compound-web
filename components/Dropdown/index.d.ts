export { Dropdown } from './Dropdown';
//# sourceMappingURL=index.d.ts.map