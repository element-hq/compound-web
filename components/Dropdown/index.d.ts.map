{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../src/components/Dropdown/index.ts"],"names":[],"mappings":"AAOA,OAAO,EAAE,QAAQ,EAAE,MAAM,YAAY,CAAC"}