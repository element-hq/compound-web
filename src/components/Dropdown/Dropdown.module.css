/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.container {
  display: flex;
  flex-direction: column;

  label {
    font: var(--cpd-font-body-md-medium);
    margin-block-end: var(--cpd-space-1x);
  }

  button {
    inline-size: 100%;
    border: 1px solid var(--cpd-color-border-interactive-primary);
    background: var(--cpd-color-bg-canvas-default);
    border-radius: 0.5rem;
    padding: var(--cpd-space-3x) var(--cpd-space-3x) var(--cpd-space-3x)
      var(--cpd-space-4x);
    box-sizing: border-box;
    color: var(--cpd-color-text-primary);
    font: var(--cpd-font-body-md-regular);
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    align-items: center;
    gap: var(--cpd-space-4x);

    svg {
      transition: transform 0.1s linear;
    }
  }

  /**
   * When the dropdown is open, rotate the arrow icon
   */
  button[aria-expanded="true"] {
    svg {
      transform: rotate(180deg);
    }
  }

  button.placeholder {
    color: var(--cpd-color-text-secondary);
  }

  .border {
    display: none;
    border-inline-start: 1px solid var(--cpd-color-border-interactive-secondary);
    border-inline-end: 1px solid var(--cpd-color-border-interactive-secondary);
    block-size: var(--cpd-space-1x);
    margin-block-start: calc(var(--cpd-space-1x) * -1);
    box-sizing: border-box;
  }

  .content {
    display: none;
    position: relative;

    ul {
      /**
       * To make the component going over the other elements
       */
      position: absolute;
      display: block;
      inline-size: 100%;
      background: var(--cpd-color-bg-canvas-default);
      border: 1px solid var(--cpd-color-border-interactive-secondary);
      border-block-start: 0;
      border-end-start-radius: var(--cpd-space-4x);
      border-end-end-radius: var(--cpd-space-4x);
      box-sizing: border-box;
      box-shadow: 0 4px 24px 0 rgb(27 29 34 / 10%);
      margin: 0;
      padding: 0;
      padding-block-end: var(--cpd-space-4x);
      cursor: pointer;

      li {
        list-style: none;
        font: var(--cpd-font-body-md-medium);
        padding: var(--cpd-space-3x) var(--cpd-space-4x);
        border-block-end: 1px solid var(--cpd-color-gray-300);
        color: var(--cpd-color-text-secondary);
        display: flex;
        justify-content: space-between;
        align-items: center;
        gap: var(--cpd-space-4x);

        @media (hover) {
          &:hover {
            background: var(--cpd-color-gray-200);
          }
        }

        &[aria-selected="true"] {
          color: var(--cpd-color-text-primary);
          background: var(--cpd-color-gray-300);
        }
      }
    }
  }

  .open {
    display: block;
  }

  .help {
    font: var(--cpd-font-body-sm-regular);
    color: var(--cpd-color-text-secondary);
  }

  .error {
    font: var(--cpd-font-body-sm-medium);
    color: var(--cpd-color-text-critical-primary);
    display: flex;
    gap: var(--cpd-space-2x);
  }

  .error,
  .help {
    margin-block-start: var(--cpd-space-2x);
  }

  &[aria-invalid="true"] {
    label {
      color: var(--cpd-color-text-critical-primary);
    }

    button {
      border-color: var(--cpd-color-text-critical-primary);
    }
  }
}
