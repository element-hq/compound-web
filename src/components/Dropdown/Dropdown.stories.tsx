/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/
import { Dropdown } from "./Dropdown";
import { fn } from "@storybook/test";
import { Meta, StoryObj } from "@storybook/react";
import { ComponentProps } from "react";

const meta = {
  title: "Dropdown",
  component: Dropdown,
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: ["defaultValue", "placeholder", "error"],
    },
  },
  argTypes: {
    label: {
      type: "string",
    },
    error: {
      type: "string",
    },
    placeholder: {
      type: "string",
    },
    values: {
      type: "string",
    },
  },
  args: {
    label: "Label",
    placeholder: "Select an option",
    onValueChange: fn(),
    values: [
      ["Option1", "Option 1"],
      ["Option2", "Option 2"],
      ["Option3", "Option 3"],
    ],
  },
} satisfies Meta<ComponentProps<typeof Dropdown>>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?node-id=799-5732&t=g2Ex9sbzgku1nTIN-4",
    },
  },
};

export const WithHelpLabel: Story = {
  args: {
    helpLabel: "Optional help text.",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?node-id=799-345&t=g2Ex9sbzgku1nTIN-4",
    },
  },
};

export const WithError: Story = {
  args: {
    error: "Select an option",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?node-id=799-370&t=g2Ex9sbzgku1nTIN-4",
    },
  },
};

export const WithDefaultValue: Story = {
  args: {
    defaultValue: "Option2",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?node-id=799-381&t=g2Ex9sbzgku1nTIN-4",
    },
  },
};
