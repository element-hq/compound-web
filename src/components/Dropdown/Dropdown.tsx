/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import ChevronDown from "@vector-im/compound-design-tokens/assets/web/icons/chevron-down";
import Check from "@vector-im/compound-design-tokens/assets/web/icons/check";
import Error from "@vector-im/compound-design-tokens/assets/web/icons/error";

import React, {
  Dispatch,
  forwardRef,
  HTMLProps,
  memo,
  RefObject,
  SetStateAction,
  useCallback,
  useEffect,
  useRef,
  useState,
  KeyboardEvent,
  useMemo,
} from "react";

import classNames from "classnames";

import styles from "./Dropdown.module.css";
import { useId } from "@floating-ui/react";

type DropdownProps = {
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The controlled value of the dropdown.
   */
  value?: string;
  /**
   * The default value of the dropdown, used when uncontrolled.
   */
  defaultValue?: string;
  /**
   * The values of the dropdown.
   * [value, text]
   */
  values: [string, string][];
  /**
   * The placeholder text.
   */
  placeholder: string;
  /**
   * The label to display at the top of the dropdown
   */
  label: string;
  /**
   * The help label to display at the bottom of the dropdown
   */
  helpLabel?: string;
  /**
   * Callback for when the value changes.
   * @param value
   */
  onValueChange?: (value: string) => void;
  /**
   * The error message to display.
   */
  error?: string;
};

/**
 * The dropdown content.
 */
export const Dropdown = forwardRef<HTMLButtonElement, DropdownProps>(
  function Dropdown(
    {
      className,
      label,
      placeholder,
      helpLabel,
      onValueChange,
      error,
      value: controlledValue,
      defaultValue,
      values,
      ...props
    },
    ref,
  ) {
    const [uncontrolledValue, setUncontrolledValue] = useState(defaultValue);
    const value = controlledValue ?? uncontrolledValue;
    const text = useMemo(
      () =>
        value === undefined
          ? placeholder
          : (values.find(([v]) => v === value)?.[1] ?? placeholder),
      [value, values, placeholder],
    );

    const setValue = useCallback(
      (value: string) => {
        setUncontrolledValue(value);
        onValueChange?.(value);
      },
      [setUncontrolledValue, onValueChange],
    );

    const [open, setOpen, dropdownRef] = useOpen();
    const { listRef, onComboboxKeyDown, onOptionKeyDown } = useKeyboardShortcut(
      open,
      setOpen,
      setValue,
    );

    const buttonRef = useRef<HTMLButtonElement | null>(null);
    useEffect(() => {
      // Focus the button when the value is set
      // Test if the value is undefined to avoid focusing on the first render
      if (value !== undefined) buttonRef.current?.focus();
    }, [value]);

    const hasPlaceholder = text === placeholder;
    const buttonClasses = classNames({
      [styles.placeholder]: hasPlaceholder,
    });
    const borderClasses = classNames(styles.border, {
      [styles.open]: open,
    });
    const contentClasses = classNames(styles.content, {
      [styles.open]: open,
    });

    /**
     * Ids for accessibility.
     */
    const labelId = useId();
    const contentId = useId();

    return (
      <div
        ref={dropdownRef}
        className={classNames(className, styles.container)}
        aria-invalid={Boolean(error)}
      >
        <label id={labelId}>{label}</label>
        <button
          className={buttonClasses}
          role="combobox"
          aria-haspopup="listbox"
          aria-labelledby={labelId}
          aria-controls={contentId}
          aria-expanded={open}
          ref={(element) => {
            // Private ref to focus the button
            buttonRef.current = element;
            // Handle forwarded ref
            if (typeof ref === "function") {
              ref(element);
            } else if (ref) {
              ref.current = element;
            }
          }}
          onClick={() => setOpen((_open) => !_open)}
          onKeyDown={onComboboxKeyDown}
          {...props}
        >
          {text}
          <ChevronDown width="24" height="24" />
        </button>
        <div className={borderClasses} />
        <div className={contentClasses}>
          <ul
            ref={listRef}
            id={contentId}
            role="listbox"
            className={styles.content}
          >
            {values.map(([v, text]) => (
              <DropdownItem
                key={v}
                isDisplayed={open}
                isSelected={value === v}
                onClick={() => {
                  setOpen(false);
                  setValue(v);
                }}
                onKeyDown={(e) => onOptionKeyDown(e, v)}
              >
                {text}
              </DropdownItem>
            ))}
          </ul>
        </div>
        {!error && helpLabel && (
          <span className={styles.help}>{helpLabel}</span>
        )}
        {error && (
          <span className={styles.error}>
            <Error width="20" height="20" />
            {error}
          </span>
        )}
      </div>
    );
  },
);

type DropdownItemProps = HTMLProps<HTMLLIElement> & {
  /**
   * Whether the dropdown item is selected.
   */
  isSelected: boolean;
  /**
   * Whether the dropdown item is displayed.
   */
  isDisplayed: boolean;
  /**
   * The text to display in the dropdown item.
   */
  children: string;
};

/**
 * A dropdown item component.
 */
const DropdownItem = memo(function DropdownItem({
  children,
  isSelected,
  isDisplayed,
  ...props
}: DropdownItemProps) {
  const ref = useRef<HTMLLIElement>(null);

  // Focus the item if the dropdown is open and the item is already selected
  useEffect(() => {
    if (isSelected && isDisplayed) {
      ref.current?.focus();
    }
  }, [isSelected, isDisplayed]);

  return (
    <li
      tabIndex={0}
      role="option"
      ref={ref}
      aria-selected={isSelected}
      {...props}
    >
      {children} {isSelected && <Check width="20" height="20" />}
    </li>
  );
});

/**
 * A hook to manage the open state of the dropdown.
 */
function useOpen(): [
  boolean,
  Dispatch<SetStateAction<boolean>>,
  RefObject<HTMLDivElement | null>,
] {
  const [open, setOpen] = useState(false);
  const ref = useRef<HTMLDivElement | null>(null);

  // If the user clicks outside the dropdown, close it
  useEffect(() => {
    const closeIfOutside = (e: MouseEvent) => {
      if (ref.current && !ref.current.contains(e.target as Node)) {
        setOpen(false);
      }
    };

    document.addEventListener("click", closeIfOutside);
    return () => document.removeEventListener("click", closeIfOutside);
  }, [setOpen]);

  return [open, setOpen, ref];
}

/**
 * A hook to manage the keyboard shortcuts of the dropdown.
 * @param open - the dropdown open state.
 * @param setOpen - the dropdown open state setter.
 * @param setValue - set the selected value and text
 */
function useKeyboardShortcut(
  open: boolean,
  setOpen: Dispatch<SetStateAction<boolean>>,
  setValue: (value: string) => void,
) {
  const listRef = useRef<HTMLUListElement>(null);
  const onComboboxKeyDown = useCallback(
    ({ key }: KeyboardEvent) => {
      switch (key) {
        // Enter and Space already managed because it's a button
        case "Escape":
          setOpen(false);
          break;
        case "ArrowDown":
          setOpen(true);
          // If open, focus the first element
          if (open) {
            (listRef.current?.firstElementChild as HTMLElement)?.focus();
          }
          break;
        case "ArrowUp":
          setOpen(true);
          break;
        case "Home": {
          setOpen(true);
          // Wait for the dropdown to be opened
          Promise.resolve().then(() => {
            (listRef.current?.firstElementChild as HTMLElement)?.focus();
          });
          break;
        }
        case "End": {
          setOpen(true);
          // Wait for the dropdown to be opened
          Promise.resolve().then(() => {
            (listRef.current?.lastElementChild as HTMLElement)?.focus();
          });
          break;
        }
      }
    },
    [listRef, open, setOpen],
  );

  const onOptionKeyDown = useCallback(
    (evt: KeyboardEvent, value: string) => {
      const { key, altKey } = evt;
      evt.stopPropagation();
      evt.preventDefault();

      switch (key) {
        case "Enter":
        case " ": {
          setValue(value);
          setOpen(false);
          break;
        }
        case "Tab":
        case "Escape":
          setOpen(false);
          break;
        case "ArrowDown": {
          const currentFocus = document.activeElement;
          if (listRef.current?.contains(currentFocus) && currentFocus) {
            (currentFocus.nextElementSibling as HTMLElement)?.focus();
          }
          break;
        }
        case "ArrowUp": {
          if (altKey) {
            setValue(value);
            setOpen(false);
          } else {
            const currentFocus = document.activeElement;
            if (listRef.current?.contains(currentFocus) && currentFocus) {
              (currentFocus.previousElementSibling as HTMLElement)?.focus();
            }
          }
          break;
        }
        case "Home": {
          (listRef.current?.firstElementChild as HTMLElement)?.focus();
          break;
        }
        case "End": {
          (listRef.current?.lastElementChild as HTMLElement)?.focus();
          break;
        }
      }
    },
    [listRef, setValue, setOpen],
  );

  return { listRef, onComboboxKeyDown, onOptionKeyDown };
}
