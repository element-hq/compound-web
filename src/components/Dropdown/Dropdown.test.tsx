/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, expect, it } from "vitest";
import { composeStories } from "@storybook/react";
import * as stories from "./Dropdown.stories";
import { render, screen } from "@testing-library/react";
import React, { type FC, useMemo, useState } from "react";
import { Dropdown } from "./Dropdown";
import userEvent from "@testing-library/user-event";

const { Default, WithHelpLabel, WithError, WithDefaultValue } =
  composeStories(stories);

const ControlledDropdown: FC = () => {
  const [value, setValue] = useState("1");
  const values = useMemo<[string, string][]>(
    () => [
      ["1", "Option 1"],
      ["2", "Option 2"],
    ],
    [],
  );
  return (
    <Dropdown
      value={value}
      onValueChange={setValue}
      values={values}
      placeholder=""
      label="Label"
    />
  );
};

describe("Dropdown", () => {
  it("renders a Default dropdown", () => {
    const { container } = render(<Default />);
    expect(container).toMatchSnapshot();
  });
  it("renders a dropdown with a help label", () => {
    const { container } = render(<WithHelpLabel />);
    expect(container).toMatchSnapshot();
  });
  it("renders a dropdown with an error ", () => {
    const { container } = render(<WithError />);
    expect(container).toMatchSnapshot();
  });
  it("renders a dropdown with a default value ", () => {
    const { container } = render(<WithDefaultValue />);
    expect(container).toMatchSnapshot();
  });
  it("can be opened", async () => {
    const user = userEvent.setup();
    const { getByRole, container } = render(<Default />);
    await user.click(getByRole("combobox"));
    expect(container).toMatchSnapshot();
  });
  it("can select a value", async () => {
    const user = userEvent.setup();
    const { getByRole, container } = render(<Default />);
    await user.click(getByRole("combobox"));

    expect(getByRole("option", { name: "Option 2" })).toBeVisible();

    await user.click(getByRole("option", { name: "Option 2" }));

    expect(getByRole("combobox")).toHaveTextContent("Option 2");

    await user.click(getByRole("combobox"));

    expect(getByRole("option", { name: "Option 2" })).toHaveAttribute(
      "aria-selected",
      "true",
    );

    // Option 2 should be selected
    expect(container).toMatchSnapshot();
  });
  it("can use keyboard shortcuts", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Default />);

    // arrowdown seems to already select Option 1... in real browsers this
    // doesn't happen. Maybe it's a user-event thing? arrowup just opens the
    // dropdown as we would expect.
    await user.type(getByRole("combobox"), "{arrowup}");
    expect(getByRole("combobox")).toHaveAttribute("aria-expanded", "true");

    await user.keyboard("{arrowdown}");
    expect(getByRole("option", { name: "Option 1" })).toHaveFocus();

    await user.keyboard("{End}");
    expect(getByRole("option", { name: "Option 3" })).toHaveFocus();

    await user.keyboard("{Enter}");

    expect(getByRole("combobox")).toHaveTextContent("Option 3");
    expect(getByRole("combobox")).toHaveAttribute("aria-expanded", "false");
  });
  it("supports controlled operation", async () => {
    const user = userEvent.setup();
    render(<ControlledDropdown />);

    expect(screen.getByRole("option", { name: "Option 1" })).toHaveAttribute(
      "aria-selected",
      "true",
    );
    await user.click(screen.getByRole("option", { name: "Option 2" }));
    expect(screen.getByRole("option", { name: "Option 2" })).toHaveAttribute(
      "aria-selected",
      "true",
    );
  });
});
