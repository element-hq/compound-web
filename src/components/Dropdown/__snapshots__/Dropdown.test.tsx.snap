// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown > can be opened 1`] = `
<div>
  <div
    aria-invalid="false"
    class="_container_87c1d6"
  >
    <label
      id="«r8»"
    >
      Label
    </label>
    <button
      aria-controls="«r9»"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-labelledby="«r8»"
      class="_placeholder_87c1d6"
      role="combobox"
    >
      Select an option
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6 _open_87c1d6"
    />
    <div
      class="_content_87c1d6 _open_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«r9»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 2
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown > can select a value 1`] = `
<div>
  <div
    aria-invalid="false"
    class="_container_87c1d6"
  >
    <label
      id="«ra»"
    >
      Label
    </label>
    <button
      aria-controls="«rb»"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-labelledby="«ra»"
      class=""
      role="combobox"
    >
      Option 2
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6 _open_87c1d6"
    />
    <div
      class="_content_87c1d6 _open_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«rb»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="true"
          role="option"
          tabindex="0"
        >
          Option 2
           
          <svg
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.55 17.575q-.2 0-.375-.062a.9.9 0 0 1-.325-.213L4.55 13q-.274-.274-.262-.713.012-.437.287-.712a.95.95 0 0 1 .7-.275q.425 0 .7.275L9.55 15.15l8.475-8.475q.274-.275.713-.275.437 0 .712.275.275.274.275.713 0 .437-.275.712l-9.2 9.2q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
            />
          </svg>
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown > renders a Default dropdown 1`] = `
<div>
  <div
    aria-invalid="false"
    class="_container_87c1d6"
  >
    <label
      id="«r0»"
    >
      Label
    </label>
    <button
      aria-controls="«r1»"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="«r0»"
      class="_placeholder_87c1d6"
      role="combobox"
    >
      Select an option
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6"
    />
    <div
      class="_content_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«r1»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 2
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown > renders a dropdown with a default value  1`] = `
<div>
  <div
    aria-invalid="false"
    class="_container_87c1d6"
  >
    <label
      id="«r6»"
    >
      Label
    </label>
    <button
      aria-controls="«r7»"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="«r6»"
      class=""
      role="combobox"
    >
      Option 2
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6"
    />
    <div
      class="_content_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«r7»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="true"
          role="option"
          tabindex="0"
        >
          Option 2
           
          <svg
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.55 17.575q-.2 0-.375-.062a.9.9 0 0 1-.325-.213L4.55 13q-.274-.274-.262-.713.012-.437.287-.712a.95.95 0 0 1 .7-.275q.425 0 .7.275L9.55 15.15l8.475-8.475q.274-.275.713-.275.437 0 .712.275.275.274.275.713 0 .437-.275.712l-9.2 9.2q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
            />
          </svg>
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown > renders a dropdown with a help label 1`] = `
<div>
  <div
    aria-invalid="false"
    class="_container_87c1d6"
  >
    <label
      id="«r2»"
    >
      Label
    </label>
    <button
      aria-controls="«r3»"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="«r2»"
      class="_placeholder_87c1d6"
      role="combobox"
    >
      Select an option
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6"
    />
    <div
      class="_content_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«r3»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 2
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
    <span
      class="_help_87c1d6"
    >
      Optional help text.
    </span>
  </div>
</div>
`;

exports[`Dropdown > renders a dropdown with an error  1`] = `
<div>
  <div
    aria-invalid="true"
    class="_container_87c1d6"
  >
    <label
      id="«r4»"
    >
      Label
    </label>
    <button
      aria-controls="«r5»"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="«r4»"
      class="_placeholder_87c1d6"
      role="combobox"
    >
      Select an option
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 14.95q-.2 0-.375-.062a.9.9 0 0 1-.325-.213l-4.6-4.6a.95.95 0 0 1-.275-.7q0-.425.275-.7a.95.95 0 0 1 .7-.275q.425 0 .7.275l3.9 3.9 3.9-3.9a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7l-4.6 4.6q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
        />
      </svg>
    </button>
    <div
      class="_border_87c1d6"
    />
    <div
      class="_content_87c1d6"
    >
      <ul
        class="_content_87c1d6"
        id="«r5»"
        role="listbox"
      >
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 1
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 2
           
        </li>
        <li
          aria-selected="false"
          role="option"
          tabindex="0"
        >
          Option 3
           
        </li>
      </ul>
    </div>
    <span
      class="_error_87c1d6"
    >
      <svg
        fill="currentColor"
        height="20"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 17q.424 0 .713-.288A.97.97 0 0 0 13 16a.97.97 0 0 0-.287-.713A.97.97 0 0 0 12 15a.97.97 0 0 0-.713.287A.97.97 0 0 0 11 16q0 .424.287.712.288.288.713.288m0-4q.424 0 .713-.287A.97.97 0 0 0 13 12V8a.97.97 0 0 0-.287-.713A.97.97 0 0 0 12 7a.97.97 0 0 0-.713.287A.97.97 0 0 0 11 8v4q0 .424.287.713.288.287.713.287m0 9a9.7 9.7 0 0 1-3.9-.788 10.1 10.1 0 0 1-3.175-2.137q-1.35-1.35-2.137-3.175A9.7 9.7 0 0 1 2 12q0-2.075.788-3.9a10.1 10.1 0 0 1 2.137-3.175q1.35-1.35 3.175-2.137A9.7 9.7 0 0 1 12 2q2.075 0 3.9.788a10.1 10.1 0 0 1 3.175 2.137q1.35 1.35 2.137 3.175A9.7 9.7 0 0 1 22 12a9.7 9.7 0 0 1-.788 3.9 10.1 10.1 0 0 1-2.137 3.175q-1.35 1.35-3.175 2.137A9.7 9.7 0 0 1 12 22"
        />
      </svg>
      Select an option
    </span>
  </div>
</div>
`;
